<a href="https://colab.research.google.com/github/YBilodeau/Sales-Predictions/blob/main/Pandas_Notes.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Pandas**

Notes transcribed from [Coding Dojo](https://www.codingdojo.com/). 


## **Mount Data | Import Libraries | Load Your Dataset**

### **Mount data**

- You need to mount Google Drive if you want to load data. 

In [ ]:
# You need to do this if you want to load data from Google Drive
# If you need help, see the Load data section
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### **Import Libraries**

- To use the pandas library, you need to import it.

In [ ]:
import pandas as pd

### **Load the Dataset**

- This is the same dataset we used last module. You can re-download the data [here](https://drive.google.com/file/d/1n48pIgKntPLbZ0N456F46qMR8S4A3shF/view) if needed.

In [ ]:
# Keep in mind that the filename will likely be different on your computer (especially later in course)
filename = '/content/drive/MyDrive/Colab Notebooks/CodingDojo/000 Data Files/bostonHousing1978.csv'
df = pd.read_csv(filename)
df

RM  LSTAT  PTRATIO  target
0    6.575   4.98     15.3    24.0
1    6.421   9.14     17.8    21.6
2    7.185   4.03     17.8    34.7
3    6.998   2.94     18.7    33.4
4    7.147   5.33     18.7    36.2
..     ...    ...      ...     ...
501  6.593   9.67     21.0    22.4
502  6.120   9.08     21.0    20.6
503  6.976   5.64     21.0    23.9
504  6.794   6.48     21.0    22.0
505  6.030   7.88     21.0    11.9

[506 rows x 4 columns]

## **Basic Operations**

### **Viewing the first and last 5 rows**

The head() method is used to view the first 5 rows. This is useful to get a quick glance/overview of what your dataframe looks like. If you want to change the number of rows that is displayed, just specify how many you would like displayed inside the method: for example, df.head(3) would display the first 3 rows of the dataframe.

In [ ]:
# Select top N number of records (default = 5)
df.head()

RM  LSTAT  PTRATIO  target
0  6.575   4.98     15.3    24.0
1  6.421   9.14     17.8    21.6
2  7.185   4.03     17.8    34.7
3  6.998   2.94     18.7    33.4
4  7.147   5.33     18.7    36.2

In [ ]:
# Select bottom N number of records (default = 5)
df.tail()

RM  LSTAT  PTRATIO  target
501  6.593   9.67     21.0    22.4
502  6.120   9.08     21.0    20.6
503  6.976   5.64     21.0    23.9
504  6.794   6.48     21.0    22.0
505  6.030   7.88     21.0    11.9

### **Check the column data types**

We want to view the column data types to see what type of data we are working with. Once we start building models, we need all of our data to be some type of numeric column (int or float) - so pay extra attention to any column of type "object."

In [ ]:
# Check the column data types using the dtypes attribute
df.dtypes

RM         float64
LSTAT      float64
PTRATIO    float64
target     float64
dtype: object

In [ ]:
# Use the shape attribute to get the number of rows and columns in your dataframe
df.shape

(506, 4)

In [ ]:
# The info method gives the column datatypes + number of non-null values
# Keep in this in mind in case you ever want to want to check for missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   RM       506 non-null    float64
 1   LSTAT    506 non-null    float64
 2   PTRATIO  506 non-null    float64
 3   target   506 non-null    float64
dtypes: float64(4)
memory usage: 15.9 KB


## **Slicing**

### **Select columns using brackets**

With square brackets, you can select one or more columns. Single square brackets will return a Pandas Series, while double square brackets will return a Pandas dataframe.

In [ ]:
# Select one column using double brackets
df[['RM']].head()

RM
0  6.575
1  6.421
2  7.185
3  6.998
4  7.147

In [ ]:
# Select multiple columns using double brackets
df[['RM', 'target']].head()

RM  target
0  6.575    24.0
1  6.421    21.6
2  7.185    34.7
3  6.998    33.4
4  7.147    36.2

In [ ]:
# This is a Pandas DataFrame
df[['RM', 'LSTAT', 'PTRATIO', 'target']].tail()

RM  LSTAT  PTRATIO  target
501  6.593   9.67     21.0    22.4
502  6.120   9.08     21.0    20.6
503  6.976   5.64     21.0    23.9
504  6.794   6.48     21.0    22.0
505  6.030   7.88     21.0    11.9

In [ ]:
# Select one column using single brackets
# This produces a pandas series which is a one-dimensional array of indexed data
df['RM'].head()

0    6.575
1    6.421
2    7.185
3    6.998
4    7.147
Name: RM, dtype: float64

In [ ]:
# This is a pandas series
type(df['RM'].head())

pandas.core.series.Series

With a pandas series, we can select rows using slicing like this: series[start_index:end_index]

The end_index is not inclusive. This behavior is very similar to Python lists.

In [ ]:
df['RM']

0      6.575
1      6.421
2      7.185
3      6.998
4      7.147
       ...  
501    6.593
502    6.120
503    6.976
504    6.794
505    6.030
Name: RM, Length: 506, dtype: float64

In [ ]:
df['RM'][0:10]

0    6.575
1    6.421
2    7.185
3    6.998
4    7.147
5    6.430
6    6.012
7    6.172
8    5.631
9    6.004
Name: RM, dtype: float64

You can also use dot notation to access a column, however it is recommended that you use the square bracket notation as much as possible.

In [ ]:
# Select column using dot notation. 
# This is not recommended.
df.RM.head()

0    6.575
1    6.421
2    7.185
3    6.998
4    7.147
Name: RM, dtype: float64

If the column name had a space in it, then the dot notation would fail. Dot notation also fails if your column has the same name of a DataFrame's attributes or methods.

### **Selecting columns using loc**

You can also use .loc to select columns, index, and slice your data. loc is often used as an alternative to selecting columns using brackets because it is more flexible as you will see during this module. 

In [ ]:
# pandas dataframe
df.loc[:, ['RM']].head()

RM
0  6.575
1  6.421
2  7.185
3  6.998
4  7.147

## **Filtering**

The dataset used in this section are for 4 different mortgages. A mortgage can be either 30 year or 15 year and have different interest rates. Later in the course, you will have to look at data documentation and explore data on your own before you do something with it. The goal of this section is to show you how to filter out data for later analysis. 

You can download the data [here](https://drive.google.com/file/d/1sItpPOF8IIsOht6a2ZpkdDxcqn-UU-cA/view).

In [ ]:
# Keep in mind that the filename will likely be different on your computer (especially later in course)
filename = '/content/drive/MyDrive/Colab Notebooks/CodingDojo/000 Data Files/mortgages.csv'
df = pd.read_csv(filename)

### **Filtering**

In [ ]:
df.head()

Month  Starting Balance  Repayment  ...  New Balance  Mortgage Name  Interest Rate
0      1         400000.00    1686.42  ...    399313.58        30 Year           0.03
1      2         399313.58    1686.42  ...    398625.44        30 Year           0.03
2      3         398625.44    1686.42  ...    397935.58        30 Year           0.03
3      4         397935.58    1686.42  ...    397243.99        30 Year           0.03
4      5         397243.99    1686.42  ...    396550.67        30 Year           0.03

[5 rows x 8 columns]

### **Filter by Mortgage Name**

As a reminder, here are some comparison operators that we may need to use to filter out the data that we want or don't want:

|Comparison Operator | Meaning |
| :----------: | :--------- |
| < | less than |
| <= | less than or equal to |
| >	| greater than |
| >= | greater than or equal to |
| == | equal |
| != | not equal |

In [ ]:
# Let's first start by looking at the values contained in the Mortgage Name column. 
# Also, I encourage you to lookup what the value_counts method does using the python help function
df['Mortgage Name'].value_counts()

30 Year    720
15 Year    360
Name: Mortgage Name, dtype: int64

In [ ]:
# Notice that the filter produces a pandas series of True and False values
mortgage_filter = df['Mortgage Name']=='30 Year'
mortgage_filter.head()

0    True
1    True
2    True
3    True
4    True
Name: Mortgage Name, dtype: bool

In [ ]:
# Approach 1 using square brackets
# Filter dataframe to get a DataFrame of only '30 Year'
df[mortgage_filter].head()

Month  Starting Balance  Repayment  ...  New Balance  Mortgage Name  Interest Rate
0      1         400000.00    1686.42  ...    399313.58        30 Year           0.03
1      2         399313.58    1686.42  ...    398625.44        30 Year           0.03
2      3         398625.44    1686.42  ...    397935.58        30 Year           0.03
3      4         397935.58    1686.42  ...    397243.99        30 Year           0.03
4      5         397243.99    1686.42  ...    396550.67        30 Year           0.03

[5 rows x 8 columns]

In [ ]:
# Approach 2 using loc
# Filter dataframe to get a DataFrame of only 30 year mortgages
df.loc[mortgage_filter, :].head()

Month  Starting Balance  Repayment  ...  New Balance  Mortgage Name  Interest Rate
0      1         400000.00    1686.42  ...    399313.58        30 Year           0.03
1      2         399313.58    1686.42  ...    398625.44        30 Year           0.03
2      3         398625.44    1686.42  ...    397935.58        30 Year           0.03
3      4         397935.58    1686.42  ...    397243.99        30 Year           0.03
4      5         397243.99    1686.42  ...    396550.67        30 Year           0.03

[5 rows x 8 columns]

In [ ]:
# Notice that it looks like nothing changed
# This is because we didn't update the dataframe after applying the filter
# The above output is local - we cannot use it anymore since we did not actually change/update the dataframe
# We would need to save the above code as a variable to use the filtered dataframe in future code cells
df['Mortgage Name'].value_counts()

30 Year    720
15 Year    360
Name: Mortgage Name, dtype: int64

In [ ]:
# Filter dataframe to get a DataFrame of only 30 year mortgages
# Notice we are overwriting df here to save just the 30 year mortgage data
# And now, only 30 year mortgages are in the output when we run .value_counts()
df = df.loc[mortgage_filter, :]
df['Mortgage Name'].value_counts()

30 Year    720
Name: Mortgage Name, dtype: int64

### **Filter by interest rate**

In [ ]:
# See if you can figure out what the value counts method does using the help function
df['Interest Rate'].value_counts()

0.03    360
0.05    360
Name: Interest Rate, dtype: int64

In [ ]:
# Notice that the filter produces a pandas series of True and False values
df['Interest Rate']==0.03

0       True
1       True
2       True
3       True
4       True
       ...  
715    False
716    False
717    False
718    False
719    False
Name: Interest Rate, Length: 720, dtype: bool

In [ ]:
interest_filter = df['Interest Rate']==0.03
df = df.loc[interest_filter, :]
df['Interest Rate'].value_counts(dropna = False)

0.03    360
Name: Interest Rate, dtype: int64

### **Combining Filters**

In the previous sections, we created `mortgage_filter` and `interest_filter` and used the `loc` command to filter the data by first applying the `mortgage_filter` and then the `interest_filter`. A more concise way to do this is shown below.


|Bitwise Logic Operator | Meaning |
| :-----: | :----- |
| & | and |
| \| |or |
| ^ | exclusive or |
| ~	| not |

In [ ]:
df = df.loc[mortgage_filter & interest_filter, :]
df   

Month  Starting Balance  ...  Mortgage Name  Interest Rate
0        1         400000.00  ...        30 Year           0.03
1        2         399313.58  ...        30 Year           0.03
2        3         398625.44  ...        30 Year           0.03
3        4         397935.58  ...        30 Year           0.03
4        5         397243.99  ...        30 Year           0.03
..     ...               ...  ...            ...            ...
355    356           8364.12  ...        30 Year           0.03
356    357           6698.61  ...        30 Year           0.03
357    358           5028.93  ...        30 Year           0.03
358    359           3355.08  ...        30 Year           0.03
359    360           1677.04  ...        30 Year           0.03

[360 rows x 8 columns]

## **Renaming and deleting columns**

### **Rename columns**
Here are two popular ways to rename DataFrame columns.

- **dictionary substitution:** very useful if you only want to rename a few of the columns.

- **list replacement:** requires a full list of names (in my experience, this is more error prone).

In [ ]:
# DataFrame before renaming columns
df.head()

Month  Starting Balance  Repayment  ...  New Balance  Mortgage Name  Interest Rate
0      1         400000.00    1686.42  ...    399313.58        30 Year           0.03
1      2         399313.58    1686.42  ...    398625.44        30 Year           0.03
2      3         398625.44    1686.42  ...    397935.58        30 Year           0.03
3      4         397935.58    1686.42  ...    397243.99        30 Year           0.03
4      5         397243.99    1686.42  ...    396550.67        30 Year           0.03

[5 rows x 8 columns]

In [ ]:
# Approach 1 dictionary substitution using rename method
df = df.rename(columns={'Starting Balance': 'starting_balance',
                        'Interest Paid': 'interest_paid', 
                        'Principal Paid': 'principal_paid'})
# DataFrame after renaming columns
df.head()

Month  starting_balance  Repayment  ...  New Balance  Mortgage Name  Interest Rate
0      1         400000.00    1686.42  ...    399313.58        30 Year           0.03
1      2         399313.58    1686.42  ...    398625.44        30 Year           0.03
2      3         398625.44    1686.42  ...    397935.58        30 Year           0.03
3      4         397935.58    1686.42  ...    397243.99        30 Year           0.03
4      5         397243.99    1686.42  ...    396550.67        30 Year           0.03

[5 rows x 8 columns]

In [ ]:
# Approach 2 list replacement
# Even if you want to rename a single column you need to list the rest of the columns
df.columns = ['month',
              'starting_balance',
              'repayment',
              'interest_paid',
              'principal_paid',
              'new_balance',
              'mortgage_name',
              'interest_rate']
df.head()

month  starting_balance  repayment  ...  new_balance  mortgage_name  interest_rate
0      1         400000.00    1686.42  ...    399313.58        30 Year           0.03
1      2         399313.58    1686.42  ...    398625.44        30 Year           0.03
2      3         398625.44    1686.42  ...    397935.58        30 Year           0.03
3      4         397935.58    1686.42  ...    397243.99        30 Year           0.03
4      5         397243.99    1686.42  ...    396550.67        30 Year           0.03

[5 rows x 8 columns]

### **Delete columns**
Here are two popular ways to delete DataFrame columns.

- **drop method:** more commonly used as it allows you to drop multiple columns at a time.

- **del:** less used than the drop method.

In [ ]:
# Approach 1
df = df.drop(columns=['new_balance'])
df.head()

month  starting_balance  ...  mortgage_name  interest_rate
0      1         400000.00  ...        30 Year           0.03
1      2         399313.58  ...        30 Year           0.03
2      3         398625.44  ...        30 Year           0.03
3      4         397935.58  ...        30 Year           0.03
4      5         397243.99  ...        30 Year           0.03

[5 rows x 7 columns]

In [ ]:
# Approach 2 use the del command
del df['starting_balance']
df.head()

month  repayment  interest_paid  principal_paid mortgage_name  interest_rate
0      1    1686.42        1000.00          686.42       30 Year           0.03
1      2    1686.42         998.28          688.14       30 Year           0.03
2      3    1686.42         996.56          689.86       30 Year           0.03
3      4    1686.42         994.83          691.59       30 Year           0.03
4      5    1686.42         993.10          693.32       30 Year           0.03

## **Identifying missing data**

### **Load the Dataset**
The file used in this notebook is called linear.csv. You can download it [here](https://drive.google.com/file/d/1B3AI1A7vJremxfhOQybPskmO09__KnXm/view).

In [ ]:
# Keep in mind that the filename will likely be different on your computer (especially later in course)
# The file is located here:  https://s3.amazonaws.com/General_V88/boomyeah2015/codingdojo/curriculum/content/chapter/linear.csv
filename = '/content/drive/MyDrive/Colab Notebooks/CodingDojo/000 Data Files/linear.csv'
df = pd.read_csv(filename)

### **Identifying missing data**

Values will be originally missing from a dataset or be a product of data manipulation. In pandas, missing values are typically called `NaN` or `None`.

Missing data can:

- Hint at data collection errors.
- Indicate improper conversion or manipulation.
- Actually not be considered missing. For some datasets, missing data can be listed as "zero", "false", "not applicable", "entered an empty string", among other possibilities.

This is an important subject as before you can do any sort of machine learning or even graph data, you need to deal with missing values. Missing values can lead to misinterpretation of data or errors.

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102 entries, 0 to 101
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x       102 non-null    float64
 1   y       94 non-null     float64
dtypes: float64(2)
memory usage: 1.7 KB


Two common methods to indicate where values in a DataFrame are missing are isna and isnull. They are exactly the same methods, but with different names.

In [ ]:
# Notice we have a Pandas Series of True and False values
df['y'].isna().head()

0    False
1    False
2    False
3    False
4     True
Name: y, dtype: bool

In [ ]:
y_missing = df['y'].isna()
# Looks at the rows that contain NaN for y
df.loc[y_missing,:]

x   y
4    69.887445 NaN
53   93.143070 NaN
60   95.542388 NaN
62  120.699573 NaN
64   82.657622 NaN
65  105.534175 NaN
92   37.546425 NaN
94   35.914205 NaN

In [ ]:
# Keep in mind that we can use the not operator (~) to negate the filter
# every row that doesn't have a nan is returned.
df.loc[~y_missing,:]

x           y
0      0.000000  -51.000000
1     25.000000  -12.000000
2    117.583220  134.907414
3    108.922466  134.085179
5     96.839983  114.530638
..          ...         ...
97   120.740859  133.922297
98   100.179788  126.827116
99    59.333765   50.257797
100  120.157757  144.754676
101  115.539152  120.697309

[94 rows x 2 columns]

In [ ]:
# The code counts the number of missing values
# sum() works because Booleans are a subtype of integers. 
df['y'].isna().sum()

8

In [ ]:
# the code above works very similarly as the code below
True + False + False 

1

## **Remove or fill in missing data**

### **Load the Dataset**

The file used in this notebook is called linear.csv. You can download it [here](https://drive.google.com/file/d/1B3AI1A7vJremxfhOQybPskmO09__KnXm/view) if needed.

In [ ]:
# Keep in mind that the filename will likely be different on your computer (especially later in course)
# The file is located here:  
filename = '/content/drive/MyDrive/Colab Notebooks/CodingDojo/000 Data Files/linear.csv'
df = pd.read_csv(filename)
df

x           y
0      0.000000  -51.000000
1     25.000000  -12.000000
2    117.583220  134.907414
3    108.922466  134.085179
4     69.887445         NaN
..          ...         ...
97   120.740859  133.922297
98   100.179788  126.827116
99    59.333765   50.257797
100  120.157757  144.754676
101  115.539152  120.697309

[102 rows x 2 columns]

### **Remove or Fill in Missing Data**

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102 entries, 0 to 101
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x       102 non-null    float64
 1   y       94 non-null     float64
dtypes: float64(2)
memory usage: 1.7 KB


### **Remove missing values**
You can remove missing values by using the `dropna` method.

In [ ]:
# You can drop entire rows if they contain 'any' nans in them or 'all'
# note that there is no row with index 4 
df.loc[0:10,:].dropna(how = 'any')

x           y
0     0.000000  -51.000000
1    25.000000  -12.000000
2   117.583220  134.907414
3   108.922466  134.085179
5    96.839983  114.530638
6    51.775940   31.376437
7    35.016737    8.764634
8    79.457646   73.285341
9    45.344909   18.859865
10   77.767132   72.946609

### **Filling in Missing Values**
There are a [variety of ways](https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html) to fill in missing values.

In [ ]:
# Looking at where missing data is located
df.loc[0:10, 'y']

0     -51.000000
1     -12.000000
2     134.907414
3     134.085179
4            NaN
5     114.530638
6      31.376437
7       8.764634
8      73.285341
9      18.859865
10     72.946609
Name: y, dtype: float64

In [ ]:
# Filling in the nan with a zero can be a bad idea
df.loc[0:10, 'y'].fillna(0)

0     -51.000000
1     -12.000000
2     134.907414
3     134.085179
4       0.000000
5     114.530638
6      31.376437
7       8.764634
8      73.285341
9      18.859865
10     72.946609
Name: y, dtype: float64

In [ ]:
# back fill in value
df.loc[0:10, 'y'].fillna(method='bfill')

0     -51.000000
1     -12.000000
2     134.907414
3     134.085179
4     114.530638
5     114.530638
6      31.376437
7       8.764634
8      73.285341
9      18.859865
10     72.946609
Name: y, dtype: float64

In [ ]:
# forward fill in value
df.loc[0:10, 'y'].fillna(method='ffill')

0     -51.000000
1     -12.000000
2     134.907414
3     134.085179
4     134.085179
5     114.530638
6      31.376437
7       8.764634
8      73.285341
9      18.859865
10     72.946609
Name: y, dtype: float64

In [ ]:
# linear interpolation (filling in of values)
df.loc[0:10, 'y'].interpolate(method = 'linear')

0     -51.000000
1     -12.000000
2     134.907414
3     134.085179
4     124.307909
5     114.530638
6      31.376437
7       8.764634
8      73.285341
9      18.859865
10     72.946609
Name: y, dtype: float64

## **Convert pandas DataFrames to NumPy arrays or dictionaries**

### **Load the Dataset**

The file used in this notebook is called linear.csv. You can download it [here](https://drive.google.com/file/d/1B3AI1A7vJremxfhOQybPskmO09__KnXm/view) if needed.

In [ ]:
# Keep in mind that the filename will likely be different on your computer (especially later in course)
# The file is located here:  
filename = '/content/drive/MyDrive/Colab Notebooks/CodingDojo/000 Data Files/linear.csv'
df = pd.read_csv(filename)
df

x           y
0      0.000000  -51.000000
1     25.000000  -12.000000
2    117.583220  134.907414
3    108.922466  134.085179
4     69.887445         NaN
..          ...         ...
97   120.740859  133.922297
98   100.179788  126.827116
99    59.333765   50.257797
100  120.157757  144.754676
101  115.539152  120.697309

[102 rows x 2 columns]

### **Convert pandas DataFrames to NumPy arrays**

While pandas DataFrames are acceptable inputs for a wide variety of libraries, there are occasions where you need to have your data in a NumPy array (some statistics packages prefer NumPy arrays).

In [ ]:
# Approach 1
df.to_numpy()

array([[  0.        , -51.        ],
       [ 25.        , -12.        ],
       [117.5832198 , 134.9074137 ],
       [108.9224662 , 134.0851795 ],
       [ 69.88744536,          nan],
       [ 96.83998269, 114.5306376 ],
       [ 51.77594025,  31.37643731],
       [ 35.01673734,   8.76463417],
       [ 79.45764646,  73.28534128],
       [ 45.3449093 ,  18.85986466],
       [ 77.76713229,  72.9466086 ],
       [ 55.85663802,  28.63728633],
       [ 51.8290553 ,  27.18413784],
       [ 36.07902502,  15.46895475],
       [ 80.63255196,  77.25552457],
       [124.4126644 , 147.817536  ],
       [ 38.88238694,  29.62914801],
       [ 85.24501874,  92.31571653],
       [105.4849878 , 111.063745  ],
       [ 36.82741469,   2.79748519],
       [ 98.61682009,  98.22796195],
       [102.1309415 , 117.0827211 ],
       [101.9740862 , 111.1555591 ],
       [122.5816416 , 147.0235042 ],
       [ 96.36839009,  99.62870641],
       [ 64.89044698,  47.94620302],
       [ 81.87156504,  80.51588907],
 

In [ ]:
# Approach 2
df.values

array([[  0.        , -51.        ],
       [ 25.        , -12.        ],
       [117.5832198 , 134.9074137 ],
       [108.9224662 , 134.0851795 ],
       [ 69.88744536,          nan],
       [ 96.83998269, 114.5306376 ],
       [ 51.77594025,  31.37643731],
       [ 35.01673734,   8.76463417],
       [ 79.45764646,  73.28534128],
       [ 45.3449093 ,  18.85986466],
       [ 77.76713229,  72.9466086 ],
       [ 55.85663802,  28.63728633],
       [ 51.8290553 ,  27.18413784],
       [ 36.07902502,  15.46895475],
       [ 80.63255196,  77.25552457],
       [124.4126644 , 147.817536  ],
       [ 38.88238694,  29.62914801],
       [ 85.24501874,  92.31571653],
       [105.4849878 , 111.063745  ],
       [ 36.82741469,   2.79748519],
       [ 98.61682009,  98.22796195],
       [102.1309415 , 117.0827211 ],
       [101.9740862 , 111.1555591 ],
       [122.5816416 , 147.0235042 ],
       [ 96.36839009,  99.62870641],
       [ 64.89044698,  47.94620302],
       [ 81.87156504,  80.51588907],
 

### **Convert pandas DataFrames to dictionaries**

This section is just to show that pandas allows you to convert your DataFrame to a variety of different objects. I encourage you to check out other pandas methods.

In [ ]:
df.to_dict()

{'x': {0: 0.0,
  1: 25.0,
  2: 117.58321979999998,
  3: 108.9224662,
  4: 69.88744536,
  5: 96.83998269,
  6: 51.77594025,
  7: 35.01673734,
  8: 79.45764646,
  9: 45.3449093,
  10: 77.76713229,
  11: 55.85663802,
  12: 51.82905529999999,
  13: 36.07902502,
  14: 80.63255196,
  15: 124.41266440000001,
  16: 38.88238694,
  17: 85.24501874,
  18: 105.4849878,
  19: 36.82741469,
  20: 98.61682009,
  21: 102.1309415,
  22: 101.9740862,
  23: 122.5816416,
  24: 96.36839009,
  25: 64.89044698,
  26: 81.87156504,
  27: 83.46634106,
  28: 129.9988213,
  29: 129.5934245,
  30: 71.30088195,
  31: 89.63463319,
  32: 82.74502445,
  33: 84.29777124,
  34: 93.6549342,
  35: 37.33160529,
  36: 52.16604202,
  37: 104.833943,
  38: 106.4639932,
  39: 79.51558318,
  40: 41.4999434,
  41: 113.6214443,
  42: 90.20451127,
  43: 52.56692661,
  44: 40.50463433,
  45: 43.45292157,
  46: 122.12520009999999,
  47: 55.16828222,
  48: 96.44691123,
  49: 119.27319740000002,
  50: 52.19022426,
  51: 114.8983489,
  

In [ ]:
# You can suppress output in a Colab or Jupyter Notebook by using a ;
df.to_dict();

## **Export pandas DataFrames to csv or Excel files**

The dataset used in this section are for 4 different mortgages. A mortgage can be either 30 year or 15 year and have different interest rates.

In [ ]:
# Keep in mind that the filename will likely be different on your computer (especially later in course)
filename = '/content/drive/MyDrive/Colab Notebooks/CodingDojo/000 Data Files/mortgages.csv'
df = pd.read_csv(filename)

### **Filter data**

Get 30 year mortgage payment table data with 3% apr. Assume after you filtered the data and you wanted to export the filtered dataset, this section goes over how to do that.

In [ ]:
mortgage_filter = df['Mortgage Name']=='30 Year'
interest_filter = df['Interest Rate']==0.03
df = df.loc[mortgage_filter & interest_filter, :]

### **Export pandas DataFrames to csv or Excel files**

In [ ]:
# Export DataFrame to csv File
df.to_csv(path_or_buf='oneMortgage.csv',index = False)

In [ ]:
# Export DataFrame to Excel File
df.to_excel(excel_writer='oneMortgage.xlsx',index=False)

After you export a file, I encourage you to look for the file and see that it was exported where you thought it would be and that it was the data you were interested in.

- click on the folder icon on the left pane (3rd icon)
- click 'Refresh'

You can find the files you exported below. You can right click on the individual files to download the files to your local computer if you like. 

In [ ]:
# Keep in mind that if you dont know what a methods parameters do,
# you can look them up using the help command:
# help(df.to_csv);

## **Aggregate Methods**

### **Load the Dataset**

The dataset used in this section are for 4 different mortgages. A mortgage can be either 30 year or 15 year and have different interest rates. Later in the course, you will have to look at data documentation and explore data on your own before you do something with it. The goal of this section is to show you how to filter out data for later analysis. You can re-download the data [here](https://drive.google.com/file/d/1sItpPOF8IIsOht6a2ZpkdDxcqn-UU-cA/view) if needed.

**Filter data**

Get 30 year mortgage payment table data with 3% apr.

In [ ]:
mortgage_filter = df['Mortgage Name']=='30 Year'
interest_filter = df['Interest Rate']==0.03
df = df.loc[mortgage_filter & interest_filter, :]

### **Aggregate Methods**

Summary statistics are useful to understand your dataset better.

| Aggregate Method |	Description |
| :----: | :----- |
|sum   |sum of values |
|cumsum	|cumulative sum|
|mean	|mean of values|
|median	|arithmetic median of values|
|min	|minimum|
|max	|maximum|
|mode	|mode|
|std	|unbiased standard deviation|
|var	|unbiased variance|
|quantile	|compute rank-based statistics of elements|

In [ ]:
# Using shape to make sure the filtered dataset size makes sense (that 12 months * 30 years = 360 payments
df.shape

(360, 8)

In [ ]:
# sum the values in a column
# total amount of interest paid over the course of the loan
df['Interest Paid'].sum()

207106.01

In [ ]:
# sum all the values across all columns
df.sum()

Month                                                           64980
Starting Balance                                          8.28431e+07
Repayment                                                      607111
Interest Paid                                                  207106
Principal Paid                                                 400005
New Balance                                               8.24431e+07
Mortgage Name       30 Year30 Year30 Year30 Year30 Year30 Year30 Y...
Interest Rate                                                    10.8
dtype: object

Notice that when "sum" is applied to the Mortgage Name column it just concatenates the strings together since Mortgage Name is not numeric.

### **Summary Statistics**

A convenient way to obtain multiple statistics from numerical columns is to use df.describe().  This will output the count, mean, standard deviation, min, max, and 25th, 50th, and 75th percentiles for all numeric columns. 

In [ ]:
# show summary stats for all numeric columns
df.describe()

Month  Starting Balance  ...    New Balance  Interest Rate
count  360.000000        360.000000  ...     360.000000   3.600000e+02
mean   180.500000     230119.808250  ...  229008.682722   3.000000e-02
std    104.067286     114722.017993  ...  115008.823040   1.042283e-16
min      1.000000       1677.040000  ...      -5.190000   3.000000e-02
25%     90.750000     136767.252500  ...  135422.742500   3.000000e-02
50%    180.500000     244736.870000  ...  243662.290000   3.000000e-02
75%    270.250000     331030.610000  ...  330171.765000   3.000000e-02
max    360.000000     400000.000000  ...  399313.580000   3.000000e-02

[8 rows x 7 columns]

If you prefer to have less precision in your results and to eliminate the scientific notation, you can set how many places after the decimal to include in your results by using .round().

In [ ]:
# to limit output to 2 places after the decimal
df.describe().round(2) 

Month  Starting Balance  ...  New Balance  Interest Rate
count  360.00            360.00  ...       360.00         360.00
mean   180.50         230119.81  ...    229008.68           0.03
std    104.07         114722.02  ...    115008.82           0.00
min      1.00           1677.04  ...        -5.19           0.03
25%     90.75         136767.25  ...    135422.74           0.03
50%    180.50         244736.87  ...    243662.29           0.03
75%    270.25         331030.61  ...    330171.76           0.03
max    360.00         400000.00  ...    399313.58           0.03

[8 rows x 7 columns]

As seen in the output above, by default, only numerical columns are included in the output. However, df.describe(include = 'all') will also provide the count, unique, top, and frequency for non-numeric columns.

In [ ]:
# this will include results for non-numeric columns
# this example demonstrates how obtain 4 places after the decimal
df.describe(include = 'all').round(4)

Month  Starting Balance  ...  Mortgage Name  Interest Rate
count   360.0000          360.0000  ...            360         360.00
unique       NaN               NaN  ...              1            NaN
top          NaN               NaN  ...        30 Year            NaN
freq         NaN               NaN  ...            360            NaN
mean    180.5000       230119.8083  ...            NaN           0.03
std     104.0673       114722.0180  ...            NaN           0.00
min       1.0000         1677.0400  ...            NaN           0.03
25%      90.7500       136767.2525  ...            NaN           0.03
50%     180.5000       244736.8700  ...            NaN           0.03
75%     270.2500       331030.6100  ...            NaN           0.03
max     360.0000       400000.0000  ...            NaN           0.03

[11 rows x 8 columns]

If you want to limit the summary output to just a particular column, you can specify the column as shown.

In [ ]:
# show summary stats for one column
df['Starting Balance'].describe()

count       360.000000
mean     230119.808250
std      114722.017993
min        1677.040000
25%      136767.252500
50%      244736.870000
75%      331030.610000
max      400000.000000
Name: Starting Balance, dtype: float64

## **Groupby**

### **What is groupby?**
In the aggregate methods section, you learned how to apply common summary statistics such as sum, mean, and median to an entire column of data.  Groupby allows you to efficiently compute aggregates on subsets of data. The example below demonstrates how groupby allows you to dive deeper into your analysis.

### **Example Code**
This data set contains customer satisfaction survey results for airline passengers.  It is a modified data set originally found on kaggle.com.  The modified dataset can be found [here](https://drive.google.com/file/d/1X6570Wm0vp8EkM_Civf3jN_zHJnHNxSw/view).  The satisfaction column provides each passengers overall rating of either "satisfied" or "neutral or dissatisfied".  There are subcategories such as "Food and drink" or "Seat comfort" that have numerical ratings on a scale of 1-5.   

### **Load the Dataset**

In [ ]:
# Keep in mind that the filename will be based on your path
filename = '/content/drive/MyDrive/Colab Notebooks/CodingDojo/000 Data Files/passenger_modified.csv'
df = pd.read_csv(filename)
df.head()

Gender  ...             satisfaction
0    Male  ...  neutral or dissatisfied
1    Male  ...  neutral or dissatisfied
2  Female  ...                satisfied
3  Female  ...  neutral or dissatisfied
4    Male  ...                satisfied

[5 rows x 15 columns]

### **Start Exploring**

In [ ]:
# obtain essential information about the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103793 entries, 0 to 103792
Data columns (total 15 columns):
 #   Column                      Non-Null Count   Dtype 
---  ------                      --------------   ----- 
 0   Gender                      103793 non-null  object
 1   Customer Type               103793 non-null  object
 2   Age                         103793 non-null  int64 
 3   Type of Travel              103793 non-null  object
 4   Class                       103793 non-null  object
 5   Flight Distance             103793 non-null  int64 
 6   Food and drink              103793 non-null  int64 
 7   Seat comfort                103793 non-null  int64 
 8   Inflight entertainment      103793 non-null  int64 
 9   On-board service            103793 non-null  int64 
 10  Baggage handling            103793 non-null  int64 
 11  Inflight service            103793 non-null  int64 
 12  Cleanliness                 103793 non-null  int64 
 13  Departure Delay in Minutes  1

We see that there are 103,793 rows of data, and there are no missing values.  There are numerical features such as age and flight distance. We also have categorical features such as Gender or Type of Travel.  

In the aggregates lesson, we learned that a great way to obtain common statistics for each column is to use df.describe

In [ ]:
# find summary stats for each numerical column and round to two decimal places
df.describe().round(2)

Age  Flight Distance  ...  Cleanliness  Departure Delay in Minutes
count  103793.00        103793.00  ...    103793.00                   103793.00
mean       39.39          1189.35  ...         3.29                       14.70
std        15.11           996.93  ...         1.31                       37.81
min         7.00            31.00  ...         1.00                        0.00
25%        27.00           414.00  ...         2.00                        0.00
50%        40.00           843.00  ...         3.00                        0.00
75%        51.00          1742.00  ...         4.00                       12.00
max        85.00          4983.00  ...         5.00                     1592.00

[8 rows x 10 columns]

Based just on the output above, we can start learning a lot about our data. For example, we see the flight distances ranged from 31 - 4983 miles, while the average was 1189 miles. We see that the average departure delay was 14.7 minutes, but the maximum delay was 1592 minutes (more than 1 day!) Take some time to consider other insights you can gain from this output.

### **Using Groupby with One Column**

While aggregate data that includes the entire column can provide many useful insights, we often want to explore our data in more detail.  Let's consider the average rating for "Seat comfort" across all passengers. We can find the mean in the output above, or we can limit our output to focus just on this value:

In [ ]:
# find the mean rating for seat comfort across all passengers
df['Seat comfort'].mean()

3.439904425153912

What if we want to group the results by class? This is where we will include .groupby() with the column we wish to group by in the parenthesis.

In [ ]:
# find the mean rating for seat comfort when passengers are grouped by class
df.groupby('Class')['Seat comfort'].mean()

Class
Business    3.761556
Eco         3.139008
Eco Plus    3.183679
Name: Seat comfort, dtype: float64

Now we see that the average "Seat Comfort" rating for Business class passengers is higher than the Eco or Eco Plus passengers.

### **Using Groupby with Multiple Columns**

We might want to further subdivide our results to see how the rating compares if we break the groups down by both class and gender.  We will include both columns as a list (notice the square brackets around the list) within the parenthesis as shown.

In [ ]:
# find the mean rating for seat comfort when grouped by both class and gender
df.groupby(['Class', 'Gender'])['Seat comfort'].mean()

Class     Gender
Business  Female    3.755691
          Male      3.767463
Eco       Female    3.217490
          Male      3.057182
Eco Plus  Female    3.249682
          Male      3.110547
Name: Seat comfort, dtype: float64

As you can see, groupby is a very powerful tool for analyzing your data at different levels of detail.  There are many more ways group and explore this data set! What other insights can you find?